// @ramp-network/ramp-instant-sdk/dist/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InternalSdkEventTypes = exports.InternalEventTypes = exports.WidgetEventTypes = exports.EventSeverity = exports.PurchaseStatus = exports.PaymentMethodName = void 0;
var PaymentMethodName;
(function (PaymentMethodName) {
    PaymentMethodName["MANUAL_BANK_TRANSFER"] = "MANUAL_BANK_TRANSFER";
    PaymentMethodName["AUTO_BANK_TRANSFER"] = "AUTO_BANK_TRANSFER";
    PaymentMethodName["CARD_PAYMENT"] = "CARD_PAYMENT";
    PaymentMethodName["APPLE_PAY"] = "APPLE_PAY";
})(PaymentMethodName = exports.PaymentMethodName || (exports.PaymentMethodName = {}));
var PurchaseStatus;
(function (PurchaseStatus) {
    PurchaseStatus["INITIALIZED"] = "INITIALIZED";
    PurchaseStatus["PAYMENT_STARTED"] = "PAYMENT_STARTED";
    PurchaseStatus["PAYMENT_IN_PROGRESS"] = "PAYMENT_IN_PROGRESS";
    PurchaseStatus["PAYMENT_FAILED"] = "PAYMENT_FAILED";
    PurchaseStatus["PAYMENT_EXECUTED"] = "PAYMENT_EXECUTED";
    PurchaseStatus["FIAT_RECEIVED"] = "FIAT_RECEIVED";
    PurchaseStatus["FIAT_SENT"] = "FIAT_SENT";
    PurchaseStatus["EXPIRED"] = "EXPIRED";
    PurchaseStatus["CANCELLED"] = "CANCELLED";
    PurchaseStatus["RELEASING"] = "RELEASING";
    PurchaseStatus["RELEASED"] = "RELEASED";
})(PurchaseStatus = exports.PurchaseStatus || (exports.PurchaseStatus = {}));
var EventSeverity;
(function (EventSeverity) {
    EventSeverity["VERBOSE"] = "VERBOSE";
    EventSeverity["DEBUG"] = "DEBUG";
    EventSeverity["INFO"] = "INFO";
    EventSeverity["WARNING"] = "WARNING";
    EventSeverity["ERROR"] = "ERROR";
    EventSeverity["CRITICAL"] = "CRITICAL";
})(EventSeverity = exports.EventSeverity || (exports.EventSeverity = {}));
var WidgetEventTypes;
(function (WidgetEventTypes) {
    WidgetEventTypes["WIDGET_CLOSE"] = "WIDGET_CLOSE";
    WidgetEventTypes["WIDGET_CONFIG_DONE"] = "WIDGET_CONFIG_DONE";
    WidgetEventTypes["WIDGET_CONFIG_FAILED"] = "WIDGET_CONFIG_FAILED";
    WidgetEventTypes["PURCHASE_CREATED"] = "PURCHASE_CREATED";
})(WidgetEventTypes = exports.WidgetEventTypes || (exports.WidgetEventTypes = {}));
var InternalEventTypes;
(function (InternalEventTypes) {
    InternalEventTypes["WIDGET_CLOSE_REQUEST"] = "WIDGET_CLOSE_REQUEST";
    InternalEventTypes["WIDGET_CLOSE_REQUEST_CANCELLED"] = "WIDGET_CLOSE_REQUEST_CANCELLED";
    InternalEventTypes["WIDGET_CLOSE_REQUEST_CONFIRMED"] = "WIDGET_CLOSE_REQUEST_CONFIRMED";
    InternalEventTypes["REQUEST_CRYPTO_ACCOUNT"] = "REQUEST_CRYPTO_ACCOUNT";
})(InternalEventTypes = exports.InternalEventTypes || (exports.InternalEventTypes = {}));
var InternalSdkEventTypes;
(function (InternalSdkEventTypes) {
    InternalSdkEventTypes["REQUEST_CRYPTO_ACCOUNT_RESULT"] = "REQUEST_CRYPTO_ACCOUNT_RESULT";
})(InternalSdkEventTypes = exports.InternalSdkEventTypes || (exports.InternalSdkEventTypes = {}));
