// @ubeswap/moola/build/types/factories/IUbeswapFeeOnTransferRouter__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "@ethersproject/providers";

import type { IUbeswapFeeOnTransferRouter } from "../IUbeswapFeeOnTransferRouter";

export class IUbeswapFeeOnTransferRouter__factory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IUbeswapFeeOnTransferRouter {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as IUbeswapFeeOnTransferRouter;
  }
}

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "amountIn",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amountOutMin",
        type: "uint256",
      },
      {
        internalType: "address[]",
        name: "path",
        type: "address[]",
      },
      {
        internalType: "address",
        name: "to",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "deadline",
        type: "uint256",
      },
    ],
    name: "swapExactTokensForTokensSupportingFeeOnTransferTokens",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];
